tests/src/certs.c

```diff
@@ -441,51 +441,51 @@ const size_t mbedtls_test_cas_len[] = {
 /* List of all available CA certificates in DER format */
 const unsigned char *mbedtls_test_cas_der[] = {
 #if defined(MBEDTLS_RSA_C)
-#if (defined(MBEDTLS_MD_CAN_SHA256) \
+#if (defined(MBEDTLS_SHA256_C) \
     || defined(PSA_WANT_ALG_SHA256))
     mbedtls_test_ca_crt_rsa_sha256_der,
 #endif /* PSA_WANT_ALG_SHA_256 */
-#if defined(PSA_WANT_ALG_SHA_1) || defined(MBEDTLS_MD_CAN_SHA1)
+#if defined(PSA_WANT_ALG_SHA_1) || defined(MBEDTLS_SHA1_C)
     mbedtls_test_ca_crt_rsa_sha1_der,
-#endif /* PSA_WANT_ALG_SHA_1 || MBEDTLS_MD_CAN_SHA1 */
+#endif /* PSA_WANT_ALG_SHA_1 || MBEDTLS_SHA1_C */
 #endif /* MBEDTLS_RSA_C */
-#if defined(MBEDTLS_PK_CAN_ECDSA_SOME) || defined(PSA_HAVE_ALG_SOME_ECDSA)
+#if defined(MBEDTLS_ECDSA_C) || defined(PSA_HAVE_ALG_SOME_ECDSA)
```

Comment:
The reasoning in the commit message is not correct in general, its conclusion might happen to be correct in some cases for other reasons (perhaps for hashes, I'd have to check), but at least in this particular case, the conclusion is not correct.

That's because in 3.6, PK, X.509 and TLS (except some parts of 1.3) are going to obey `USE_PSA`: if it's unset, then they're going to use the legacy version and not even look if the PSA version is available, and the other way round when it's set. So, `PK_CAN_ECDSA_SOME` is really not the same as `ECDSA_C || PSA_HAVE_ALG_SOME_ECDSA` and neither is `PK_CAN_ECDSA_SOME || PSA_HAVE_ALG_SOME_ECDSA`.

I wouldn't be surprised if that was the reason for some of the CI failures. Generally speaking, I'd say most of the `_CAN_` macros in 3.6 are there for a reason, and they're often more than just a shortcut for `legacy || psa`, so we should be think carefully when tempted to remove them in 3.6.

Reply:
Manuel's [transition guide](https://github.com/Mbed-TLS/mbedtls/blob/mbedtls-3.6/docs/architecture/psa-migration/transition-guards.md) explains the various complicated macro, and often explains why they're there.

Even for hashes, there are or have been ways to configure the library where `psa/crypto_config.h` ends up being read, so some `PSA_WANT` symbols are defined, but `MBEDTLS_PSA_CRYPTO_CLIENT` is disabled, so `PSA_WANT_FOO || MBEDTLS_FOO_C` is not equivalent to `MBEDTLS_MD_CAN_FOO` which correctly ignores `PSA_WANT_FOO` when PSA crypto is disabled.

Reply:
Hmm I see. For some reason `depends.py` fails when I keep these as the `MBEDTLS_XYZ_CAN` dependencies (even when I use major-version guards) and I have no idea why. I'll have to pick @mpg brains about this to figure out what is going wrong.

Reply:
Well that makes me feel silly. The problems were caused by me somewhere accidentally introducing the typo `PSA_WANT_ALG_SHA256` (should be `PSA_WANT_ALG_SHA_256` with an extra underscore). Using the plain `MBEDTLS_SHA256_C` macros was making it work because they would select the code block both when PSA was enabled and when the legacy SHA256 was enabled, so that part of the guard worked for both cases.

Reply:
Ah, I think we've all done that at some point. Good that the mystery is solved!